pkg/api/datasources.go

```diff
@@ -855,6 +856,10 @@ func (hs *HTTPServer) checkDatasourceHealth(c *contextmodel.ReqContext, ds *data
 		payload["details"] = jsonDetails
 	}
 
+	// treat response as downstream since it means we've got
+	// a proper response from the plugin.
+	requestmeta.WithDownstreamStatusSource(c.Req.Context())
```

Comment:
Leaving here from our 1:1 so we don't forget:

If we mutate `context.Context` here, but the error happens after we return from this function (e.g.: at a middleware-level, after the wrapped function is called) and the error does not come from Grafana, then we still have `downstream` status.

Example pseudo-code of what I mean:

```go
func doRequest(ctx context.Context) (*http.Response, error) {
	// ...
	// Mutate the context directly by marking it as downstream source
	requestmeta.WithDownstreamStatusSource(c.Req.Context())
	// do downstream response
	// ...
	return response, err
}

func processResponse(ctx context.Context, resp *http.Response) error {
	// ...
	if err != nil {
		// This error will be treated as downstream
		return fmt.Errorf("internal grafana error: %w", err)
	}
	if ... {
		// This error will be treated as downstream
		panic("grafana panic")
	}
	return nil
}

// Do downstream request
resp, err := doRequest(ctx)
if err != nil {
	return err
}

// Grafana's code
if err := processResponse(ctx, resp); err != nil {
	return err
}
```

But as of right now I'm not too sure what's the best way to solve this

Reply:
Thats a good observation! 

I think its a good trade off. In this solution we are going from ignoring 100% of the request to at least include ~80%. 
But perhaps we should set it back to `server|core` in some error handlers. ex panics. I think the next step for this idea to be fully successful we need to have a stronger contract between grafana and the datasources about the source of the error. But I really don't wanna block on that because I expect that too take much more time :) Lets not let perfection get in the way of good!